torch/utils/data/datapipes/iter/combining.py

```diff
@@ -55,12 +57,88 @@ def __iter__(self):
+        as requested by the child DataPipes.
+    """
+    def __init__(self, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        self.main_datapipe = datapipe
+        self.dp: Iterator[Any]
+        self.num_instances = num_instances
+        self.buffer: Deque = deque()
+        self.buffer_size = buffer_size
+        self.child_pointers = [0] * num_instances  # Indicate the indices of the next element to get
+        self.slowest_ptr = 0
+        self.leading_ptr = 0
+        self.end_ptr = float('inf')
+
+    def get_next(self, instance_id):
+        if not hasattr(self, "dp"):
+            self.dp = iter(self.main_datapipe)
+        while self.child_pointers[instance_id] < self.end_ptr:
+            if not self.buffer or self.child_pointers[instance_id] > self.leading_ptr:
+                self.leading_ptr = self.child_pointers[instance_id]
+                if self.leading_ptr - self.slowest_ptr > self.buffer_size:
+                    raise BufferError("ForkIterDataPipe buffer overflow," +
+                                      f"buffer size {self.buffer_size} is insufficient.")
+                try:
+                    self.buffer.append(self.dp.__next__())
+                    self.child_pointers[instance_id] += 1
+                    yield self.buffer[-1]
+                except StopIteration:
+                    self.end_ptr = self.leading_ptr
+            else:  # Child pointer is slower than or equal to the leading_ptr
+                buffer_index = self.child_pointers[instance_id] - self.slowest_ptr
+                return_val = self.buffer[buffer_index]
+                self.child_pointers[instance_id] += 1
+                if self.child_pointers[instance_id] - 1 == self.slowest_ptr:
+                    new_min = min(self.child_pointers)  # Can optimize by avoiding the call to min()
+                    if self.slowest_ptr < new_min:
+                        self.slowest_ptr = new_min
+                        self.buffer.popleft()
+                yield return_val
+
+
+class ChildDataPipe(IterDataPipe):
+    r""" :class:`ChildDataPipe`.
+
+        Iteratable Datapipe that is a child of a main DataPipe. The instance of this class
+        will pass its instance_id to get the next value from its main DataPipe.
+        args:
+            main_datapipe: Main DataPipe with a method 'get_next(instance_id)'
+            instance_id: integer identifier of this instance
+    """
+    def __init__(self, main_datapipe, instance_id: int):
+        get_next_op = getattr(main_datapipe, "get_next", None)
+        if not callable(get_next_op):
+            raise NotImplementedError("Main Datapipe must have method 'get_next' implemented.")
+        self.main_data_pipe = main_datapipe
+        self.instance_id = instance_id
+
+    def __iter__(self):
+        # TODO: Allow repeated calls to iter() and keep producing outputs, or explicitly asks user to call .fork again
```

Comment:
@VitalyFedyunin The current behavior is that, after the first read through, the subsequent calls to `iter` returns empty.

```
dp1, dp2 = input_dp.fork(2)
print(list(dp1))  # Returns [0, ..., 10]
print(list(dp1))  # Returns []
print(list(dp2))  # Returns [0, ..., 10]
```

If we want subsequent call to read from the main DataPipe again, we will have to decide how the buffer will interact with that. Or perhaps creating a separate object.

Reply:
I think the best option for such pipes (with multiple consumers) would be:
1. recreate buffer and update all positions of other consumers to zero
2. if any of the consumers haven't reached the end of buffer (at the moment of reset) - produce warning

Reply:
I implemented this as requested.